File: src/main/java/com/milaboratory/mixcr/cli/CommandExportReads.java
/*
 * Copyright (c) 2014-2021, MiLaboratories Inc. All Rights Reserved
 *
 * BEFORE DOWNLOADING AND/OR USING THE SOFTWARE, WE STRONGLY ADVISE
 * AND ASK YOU TO READ CAREFULLY LICENSE AGREEMENT AT:
 *
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 */
package com.milaboratory.mixcr.cli;

import cc.redberry.pipe.CUtils;
import com.milaboratory.core.io.sequence.SequenceRead;
import com.milaboratory.core.io.sequence.SequenceWriter;
import com.milaboratory.core.io.sequence.fastq.PairedFastqWriter;
import com.milaboratory.core.io.sequence.fastq.SingleFastqWriter;
import com.milaboratory.mixcr.basictypes.VDJCAlignments;
import com.milaboratory.mixcr.basictypes.VDJCAlignmentsReader;
import com.milaboratory.util.SmartProgressReporter;
import picocli.CommandLine.Command;
import picocli.CommandLine.Parameters;

import java.io.IOException;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

@Command(name = "exportReads",
        sortOptions = true,
        separator = " ",
        description = "Export original reads from vdjca file.")
public class CommandExportReads extends ACommandWithOutputMiXCR {
    @Parameters(description = "input.vdjca [output_R1.fastq[.gz] [output_R2.fastq[.gz]]]", arity = "1..3")
    public List<String> inOut;

    @Override
    protected List<String> getInputFiles() {
        return Collections.singletonList(inOut.get(0));
    }

    @Override
    public List<String> getOutputFiles() {
        if (inOut.size() == 1)
            return Collections.emptyList();

        if (inOut.size() == 2)
            return Collections.singletonList(inOut.get(1));

        if (inOut.size() == 3)
            return Arrays.asList(inOut.get(1), inOut.get(2));

        throwValidationException("Required parameters missing.");
        return null;
    }

    @Override
    @SuppressWarnings("unchecked")
    public void run0() throws Exception {
        try (VDJCAlignmentsReader reader = new VDJCAlignmentsReader(getInputFiles().get(0));
             SequenceWriter writer = createWriter()) {
            SmartProgressReporter.startProgressReport("Extracting reads", reader, System.err);

            for (VDJCAlignments alignments : CUtils.it(reader)) {
                List<SequenceRead> reads = alignments.getOriginalReads();
                if (reads == null)
                    throwExecutionException("VDJCA file doesn't contain original reads (perform align action with -g / --save-reads option).");

                for (SequenceRead read : reads) {
                    if (read.numberOfReads() == 1 && (writer instanceof PairedFastqWriter))
                        throwExecutionException("VDJCA file contains single-end reads, but two output files are specified.");

                    if (read.numberOfReads() == 2 && (writer instanceof SingleFastqWriter))
                        throwExecutionException("VDJCA file contains paired-end reads, but only one / no output file is specified.");

                    writer.write(read);
                }
            }
        }
    }

    public SequenceWriter<?> createWriter() throws IOException {
        List<String> outputFiles = getOutputFiles();
        switch (outputFiles.size()) {
            case 0:
                return new SingleFastqWriter(System.out);
            case 1:
                return new SingleFastqWriter(outputFiles.get(0));
            case 2:
                return new PairedFastqWriter(outputFiles.get(0), outputFiles.get(1));
        }
        throw new RuntimeException();
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/cli/CommandExportReads.kt
/*
 * Copyright (c) 2014-2023, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli

import cc.redberry.pipe.util.forEach
import com.milaboratory.app.ApplicationException
import com.milaboratory.app.InputFileType
import com.milaboratory.app.ValidationException
import com.milaboratory.app.logger
import com.milaboratory.core.io.sequence.PairedRead
import com.milaboratory.core.io.sequence.SequenceWriter
import com.milaboratory.core.io.sequence.SingleRead
import com.milaboratory.core.io.sequence.fastq.PairedFastqWriter
import com.milaboratory.core.io.sequence.fastq.SingleFastqWriter
import com.milaboratory.mixcr.basictypes.VDJCAlignmentsReader
import picocli.CommandLine.Command
import picocli.CommandLine.Parameters
import java.nio.file.Path

@Command(
    description = ["Export original reads from vdjca file."]
)
class CommandExportReads : MiXCRCommandWithOutputs() {
    @Parameters(
        description = ["Path to input file with alignments"],
        index = "0",
        paramLabel = "input.vdjca",
        arity = "1"
    )
    lateinit var input: Path

    @Parameters(
        description = [
            "Path where to write reads from input alignments.",
            "Will write to output if omitted.",
            "Will write to single file if specified only one output file.",
            "Will write paired reads to two files if specified two output files.",
        ],
        index = "1",
        paramLabel = "[output_R1.fastq[.gz] [output_R2.fastq[.gz]]]",
        hideParamSyntax = true,
        arity = "0..2"
    )
    public override val outputFiles: List<Path> = mutableListOf()

    override val inputFiles
        get() = listOf(input)

    override fun validate() {
        ValidationException.requireFileType(input, InputFileType.VDJCA)
        outputFiles.forEach { output ->
            ValidationException.requireFileType(output, InputFileType.FASTQ)
        }
    }

    override fun initialize() {
        if (outputFiles.isEmpty())
            logger.redirectSysOutToSysErr()
    }

    override fun run1() {
        VDJCAlignmentsReader(inputFiles.first()).use { reader ->
            createWriter().use { writer ->
                reader
                    .reportProgress("Extracting reads")
                    .forEach { alignments ->
                        val reads = alignments.getOriginalReads()
                            ?: throw ApplicationException("VDJCA file doesn't contain original reads (perform align action with -g / --save-reads option).")
                        for (read in reads) {
                            when (writer) {
                                is PairedFastqWriter -> {
                                    if (read.numberOfReads() == 1)
                                        throw ValidationException("VDJCA file contains single-end reads, but two output files are specified.")
                                    writer.write(read as PairedRead)
                                }

                                is SingleFastqWriter -> {
                                    if (read.numberOfReads() == 2)
                                        throw ValidationException("VDJCA file contains paired-end reads, but only one / no output file is specified.")
                                    writer.write(read as SingleRead)
                                }
                            }
                        }
                    }
            }
        }
    }

    private fun createWriter(): SequenceWriter<*> {
        val outputFiles = outputFiles
        return when (outputFiles.size) {
            0 -> SingleFastqWriter(System.out)
            1 -> SingleFastqWriter(outputFiles[0].toFile())
            2 -> PairedFastqWriter(outputFiles[0].toFile(), outputFiles[1].toFile())
            else -> throw IllegalArgumentException()
        }
    }
}
